packages/flutter_tools/lib/src/device.dart

```diff
@@ -441,3 +447,31 @@ class DiscoveredApp {
   final String id;
   final int observatoryPort;
 }
+
+// An empty device log reader
+class NoOpDeviceLogReader implements DeviceLogReader {
+  NoOpDeviceLogReader(this.name);
+
+  @override
+  final String name;
+
+  @override
+  int appPid;
+
+  @override
+  Stream<String> get logLines => const Stream<String>.empty();
+}
+
+// A portforwarder which does not support forwarding ports.
+class NoOpDevicePortForwarder implements DevicePortForwarder {
```

Comment:
I added a `NoopPortForwarder` previously for flutter-tester - the implementation is basically the same, though mine throws if you call `forward` with two different port numbers. Should we merge them? Is it safe to let this one throw if someone calls it trying to map ports too?

https://github.com/flutter/flutter/blob/789376b830d08e7a4ee27f47861d2d473eac728d/packages/flutter_tools/lib/src/tester/flutter_tester.dart#L236